### ISSUE TITLE
[css-anchor-position-1] Allow anchor references to match names in outer tree scopes

### ISSUE THREAD

**Issue by xiaochengh** [w3c/csswg-drafts#9408]
Anchor names and references are both tree scoped.

Currently, we only allow exact matches. In other words, anchor names can only be referenced in tree scopes where they are defined.

Proposal: allow anchor references in inner tree scopes to match names defined in outer tree scopes (like how we match at-rule names).

Reason:

The current way it's matched can be problematic according to feedback (@e111077), when we want elements in an inner tree scope to anchor to external anchors. For example:

```html
<div style="anchor-name: --anchor-1"></div>

<custom-element>
  #shadow-root
    <div class="anchored"></div>
    <style>
      /* Try to anchor the element against --anchor-1 but can't */
      .anchored { inset-block-start: anchor(??? top); }
    </style>
</custom-element>
```

Currently the only proper way to anchor the `.anchored` element against `--anchor-1` is to use a `::part` selector in the outer tree scope. And this won't work if `.anchored` is further nested in another shadow tree, because we don't allow `::part` selectors to chain.

If we allow inner tree scope's references to match outer tree scope's definitions, then we can simply put `anchor(--my-anchor-1 top)` in the inner tree scope's style sheets, regardless of how deep the inner scope is.

(In retrospect, we made anchor name use exact tree scope matching because we didn't know about use cases and wanted to keep implementation simple, see https://github.com/w3c/csswg-drafts/issues/7916#issuecomment-1287436800; now we have use cases)

**Comment by xiaochengh** [comment]
Minor correction: when `.anchored` is in a deep-nested tree scope, we can still style it with `::part`, but also need a long chain of [`[exportparts]`](https://drafts.csswg.org/css-shadow-parts/#exportparts-attr), which is still quite a burden to maintain.

Also, with the proposal, we can solve everything nicely without breaking encapsulation regardless of how deep the shadow tree nesting is, by passing the anchor name into the custom element via a custom property. For example:

```html
<custom-element style="--custom-element-anchor-name: --anchor-1">
  #shadow-root
    <div class="anchored"></div>
    <style>
      .anchored { inset-block-start: anchor(var(--custom-element-anchor-name) top); }
    </style>
</custom-element>
<script>
```

**Comment by e111077** [comment]
Seeing how chrome is about to ship, is there any possible way to get shadow dom support at all? This makes css anchor positioning effectively useless for web components. cc @tabatkins who touched this last for routing

**Comment by KonnorRogers** [comment]
+1 for ShadowDOM support.

we have an internal utility called `<sl-popup>` in Shoelace that uses Floating UI currently.

It supports setting an arbitrary `anchorElement` which could live in any root. As I understand currently, Anchored Region would only allow us to set anchors in the same shadow root as the anchor from the popup utility.

Without cross-root anchoring, it feels like anchored region would do very little to replace our current usage of Floating UI because it cannot accept arbitrary anchors. or at the very least, anchors from "parent" DOM trees.

**Comment by Westbrook** [comment]
This issue seems so much like [accessibility with shadow roots,](https://alice.pages.igalia.com/blog/how-shadow-dom-and-accessibility-are-in-conflict/) if it's not included by default it could be years before it's possible to get enough weight behind a productive solution. We should take care not to ship something to the web platform that doesn't support all of the web platform!

Luckily there is some prior art to lean on here.

Take for instance [`arialabelledbyelements`](https://w3c.github.io/aria/#dom-ariamixin-arialabelledbyelements) a DOM property that allows an element to be labeled by an array of element references that are in the same shadow tree or an ancestor tree. We may be waiting on long-rumored Blink and Mozilla implementations of this, but they are getting close and will do some great things to accessibility with shadow roots. Can we find a reciprocal version of this for anchor references?

Some things that I could certainly see as important to work out in this context:

**Unmanaged sharing**
I'm not quite sure how this could be possible but for anyone interested in going down this path.
- if there is some sort of sharing between trees, and it is not managed in some way, which tree "wins"? e.g. if there is an `anchor-name: --anchor-1` in the local DOM tree and an `anchor-name: --anchor-1` in the parent DOM tree, which gets leveraged?
- similar if this sharing is non-managed, would it be expected that CSS would need to walk up ALL DOM trees to _possibly_ resolve an anchor? 

I'm not sure there are good or performance answers to these questions but they should be discussed well before something in this realm hits "stable".

**Managed sharing**
What would this look like? The main goal would seem to prevent anchor name collision, but this could also manage the distance to which a browser would need to go to resolve an anchor name.
- is a CSS custom property enough to pass an anchor name across multiple shadow trees? [See example](https://github.com/w3c/csswg-drafts/issues/9408#issuecomment-1734672692).
- could an element with a shadow root define how many trees away an anchor name could be searched for?
- do we need something like downward facing [`exportparts`](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/exportparts) to make them available to child trees?
- [the reference target proposal](https://github.com/WICG/aom/blob/gh-pages/reference-target-explainer.md) may provide prior art here, as well

I look forward to seeing this make its way into the larger conversation before any "stable" API is marketed to the larger web development community as ready to go!

**Comment by vospascal** [comment]
> if it's not included by default it could be years before it's possible to get enough weight behind a productive solution. We should take care not to ship something to the web platform that doesn't support all of the web platform!

i agree on this, having to think about exceptions / and caveats or not even being able to use something because its excluded from ShadowDOM is far from ideal.

**Comment by michaelwarren1106** [comment]
another +1 for considering shadow roots for anchor positioning. 

as a larger issue, is there a recommended way that shadow root / cross root functionality can become part of the requirements of new (existing too?) proposals?

it seems like a fairly common concern that shadow roots be taken into account for new css features.

**Comment by asyncLiz** [comment]
A strong +1 to echo this requirement. Without full shadow dom support, CSS anchor roots are a no-go for [Material Web Components](https://github.com/material-components/material-web). 😞

**Comment by filimon-danopoulos** [comment]
I would be very sad to see new features that I can't use as a web component author. The design system I work with has a popout feature similar to the Shoelace one.

**Comment by astearns** [comment]
@tabatkins can you comment on your reasons for moving this issue to level 2? Is this something we could bring back in to level 1?

**Comment by trusktr** [comment]
What about new functions?

```css
.positioned-notice {
    position-anchor: crossrootnearest(--anchor-el);
}
.positioned-notice2 {
    top: anchor(crossrootupward(--anchor-el));
}
```

where here `crossrootnearest` finds the nearest matching anchor in this order:

- check the current root
- if not found, check all roots below the current root in tree order
- if not found, check the next root upward
- if not found, check all roots downward except the one we just went upward from
- if not found, repeat the previous step until found, or until there are no more upward roots

and where here `crossrootupward` finds the nearest matching anchor in this order:

- check the current root
- if not found, check the next root upward
- if not found, repeat the previous step until found, or until there are no more upward roots

**Comment by Magaav** [comment]
2025 still not a sign of parent/sibiling tree defined anchors being acessible from within inner shadow-root trees?

**Comment by justinfagnani** [comment]
A use case I have for something here is a "callout" box that has an arrow that points to the anchor.

Like so:

<img width="439" height="459" alt="Image" src="https://github.com/user-attachments/assets/43c44897-516f-4027-bfe1-f6e9fa4d1547" />

I'd like the arrow placement to be automatic, but I've only been able to do that if the arrow element is positioned directly itself, in addition to the callout's box, which means that currently the anchor has to be inside the shadow root along with the callout box and arrow.

Putting the anchor and positioned element in the shadow root means that the component user now no longer has full control over the anchor positioning with CSS. I have to invent a positioning API for them.

What I want to be able to do is get the host's anchor that was set with `position-anchor`, then use that in `anchor()` functions to place the arrow element.

**Comment by tabatkins** [comment]
Okay, as part of solving #7916 (and slightly reworking how we described #5984), this is now properly defined across Scoping and Anchor Position. Anchor references (`position-anchor`, `anchor()`, etc) have well-defined tree-crossing behavior, and can match names defined in ancestor trees.

(They still can't match names defined in lower shadows; that'll need to be a separate feature that "lifts" a name out. Need to coordinate what that'll look like across a couple of similar properties.)

**Comment by andruud** [comment]
This seems to have been edited in without a resolution backing it, deviating from the current [interoperable](https://wpt.fyi/results/css/css-anchor-position/anchor-name-in-shadow.html?label=master&label=experimental&aligned&view=interop&q=label%3Ainterop-2025-anchor-positioning) and intentional behavior.

The change makes sense to me, but we could at least take an async resolution on it. Or at the very least mark it as "accepted by editor discretion", if that's indeed what this is.

**Comment by woody-li** [comment]
Sorry to border, I has a different case about it and not sure whether to create a new issue.

In the following DOM:
```html
<x-test>
  #shadow-dom
    <slot />
  #light-dom
    <button style="anchor-name: --button-box" />
<x-test>
```

Reference failed if set position-anchor in shadow dom styles:
```css
::slotted(button)::after{
    content: '';
    display: block;
    position: absolute;
    position-anchor: --button-box;
    inset: auto anchor(right) anchor(bottom) anchor(left);
    height: 2px;
    background: black;
}
```

But works well with set in light dom:
```css
button::after{
    position-anchor: --button-box;
}
```

----

But the `::slotted(button)` also matches the light dom element, why the difference?
